hydroflow/src/util/tcp.rs

```diff
@@ -83,60 +83,65 @@ pub async fn bind_tcp<T: 'static, Codec: 'static + Clone + Decoder + Encoder<T>>
 
     let bound_endpoint = listener.local_addr()?;
 
-    let (tx_egress, mut rx_egress) = unsync_channel(None);
-    let (tx_ingress, rx_ingress) = unsync_channel(None);
-
-    let clients = Rc::new(RefCell::new(HashMap::new()));
-
-    spawn_local({
-        let clients = clients.clone();
-
-        async move {
-            while let Some((payload, addr)) = rx_egress.next().await {
-                let client = clients.borrow_mut().remove(&addr);
-
-                if let Some(mut sender) = client {
-                    let _ = SinkExt::send(&mut sender, payload).await;
-                    clients.borrow_mut().insert(addr, sender);
-                }
-            }
-        }
-    });
+    let (send_egress, mut recv_egress) = unsync_channel::<(T, SocketAddr)>(None);
+    let (send_ingres, recv_ingres) = unsync_channel(None);
 
     spawn_local(async move {
+        let send_ingress = send_ingres;
+        let mut peers_send = HashMap::new();
+        let mut peers_recv = StreamMap::new();
+
         loop {
-            let (stream, peer_addr) = if let Ok((stream, _)) = listener.accept().await {
-                if let Ok(peer_addr) = stream.peer_addr() {
-                    (stream, peer_addr)
-                } else {
-                    continue;
+            // Calling methods in a loop, futures must be cancel-safe.
+            select! {
+                // Prioritize accepting clients first, then sending outgoing messages, then receiving more messages.
+                biased;
+                // Accept new clients.
+                new_peer = listener.accept() => {
+                    let Ok((stream, _addr)) = new_peer else {
+                        continue;
+                    };
+                    let Ok(peer_addr) = stream.peer_addr() else {
+                        continue;
+                    };
+                    let (peer_send, peer_recv) = tcp_framed(stream, codec.clone());
+
+                    // TODO: Using peer_addr here as the key is a little bit sketchy.
```

Comment:
I think this merits filing an issue to be resolved in future.

Reply:
#1506